### PR TITLE
Metrics on canvas

### LINKED ISSUE #1074: Metrics on canvas
Placeholder.

### PR DESCRIPTION
**Pull request by foot**
screen shot 2016-03-03 at 13 19 01

Fixes #1074

Bugs:

- [x] Fix metric formatting, either w/ BE support from a `metric.format`, or a local mapping. It might be good to get `metric.label` here to keep the metrics consistent w/ the labels in the details panel.
- [x] Figure out what metrics are actually available to display by re-scanning `[node].metrics` onAdd/onUpdate.
- [x] Fix text size
- [x] Fix 0.0001 displaying a tiny bit of metric-area when it gets rounded away to 0.00
- [x] ~~Clarify what it means to have an agg node metric. (In metric selector perhaps. `avg`)~~ -No agg metrics for now.
- [x] ~~Do we want to show "host cpu" on the processes topo, is that interesting?~~ - No
- [x] Fixes metric keyboard sorting order.
- [x] Keep selected metric in the url state.

Features

- [x] Fade the node border color when a metric is selected to give the metric more emphasis while still (just) keeping the ability to determine a node's rank.
- [x] `<`, `>`, and `q` kb shortcuts to cycle metrics, like the `top` command.
- [x] Different color for different metrics for more easily recognising when cycling.
- [x] As soon as a metric is non-zero bump up the fill to start at effectively 10% for easier detection of 0/non-zero metrics.
- [x] Pin / unpin with by re-clicking metric.
- [x] Log scale for load (50) / open-files (100000) as these don't really have a max, though we are considering ulimit for open-files
- [x] Update metrics every 3s as the reports come in from the probe. (mocked)
- [x] Change colors away from green/red/orange.

### CODE REVIEW

**Comment by paulbellamy** [comment]
The hover on highlight with this seems to also highlight all connected nodes...

**Comment by foot** [comment]
Feature!

But will play with that. I guess its not stuff you might necessarily really be interested in, so, distracting.

**Comment by paulbellamy** [comment]
Need to figure out a better way to do the pruning for the websocket diff. Maybe `RenderableNode.Diff(other RenderableNode)`?

The js seems to go a bit wonky when the websocket updates arrive. There's a metric on every node with the value `NaN` after..

**Comment by foot** [comment]
- [ ] Fix metric formatting, either w/ BE support from a `metric.format`, or a local mapping.
- [ ] Keep a node's metrics in-line somehow, e.g. having details panel open shows the metrics in a diff state than to the canvas. Perhaps a more global store here.
- [ ] Figure out what metrics are actually available to display by re-scanning `[node].metrics` onAdd/onUpdate.

Questions
 - How often do we want the metrics to update? It sounded good to have every changed pushed onto the WS, but thats a lot of stuff perhaps.
 - How often does the server get a new value for a metric? every 15s or 3s?
 - We could do something similar to the details panel where we build a buffer up from the BE and stream it out slowly to give the (false) impression of "live" metrics.

**Comment by paulbellamy** [comment]
> How often does the server get a new value for a metric? every 15s or 3s?

Everytime there's a new report so every 3s.

**Comment by foot** [comment]
Figure out how we should handle `max`. at the mo:
- `cpu`: max is `100`. (is it 200 w/ 2 cores?)
- `load`/`memory`/`open fds`: the max value in the 15 samples the BE stores?

Does it make sense to try and figure out other values for max, e.g.
 - Total available memory on the system for memory.
 - `ulimit` for open files?

**Comment by paulbellamy** [comment]
>  (is it 200 w/ 2 cores?)

yes

> Total available memory on the system for memory. `ulimit` for open files

Probably, yeah

**Comment by foot** [comment]
Nice! #1149 would be pretty great.

I'm not sure about load.
 1. Take the global max of the topo
 2. Choose a very large value like `100` and use a log scale for the fill/bar.
 
A log scale could work for `open files` too as the max is probably something v.large.

**Comment by paulbellamy** [comment]
Upon reflection, I'm not convinced the `PruneWithMetrics` hack will work for this. Needs revisiting.

**Comment by paulbellamy** [comment]
- [ ] Figure out how to send the metric summaries in deltas and topologies updates
- [ ] Only send the last/latest value for metrics in the summaries

**Comment by foot** [comment]
rfr @davkal 

Just the JS code in this branch in the end.

File: client/app/scripts/actions/app-actions.js [hunk 1]
```diff
@@ -12,6 +13,42 @@ import AppStore from '../stores/app-store';
...
+    metricType: AppStore.getAvailableCanvasMetricsTypes()[nextMetric]
+  });
+  updateRoute();
+}
+
+export function pinMetric(metricId) {
+  AppDispatcher.dispatch({
+    type: ActionTypes.PIN_METRIC,
+    metricId,
+    metricType: AppStore.getAvailableCanvasMetricsTypes()[metricId]
```
Comment (davkal) [inline client/app/scripts/actions/app-actions.js]: Why is `metricType` computed here instead of the AppStore?

File: client/app/scripts/charts/nodes-chart.js [hunk 2]
```diff
@@ -74,10 +74,10 @@ export default class NodesChart extends React.Component {
         edges: makeMap()
       });
     }
+    //
     // FIXME add PureRenderMixin, Immutables, and move the following functions to render()
-    if (nextProps.forceRelayout || nextProps.nodes !== this.props.nodes) {
-      _.assign(state, this.updateGraphState(nextProps, state));
-    }
+    _.assign(state, this.updateGraphState(nextProps, state));
+
```
Comment (davkal) [inline client/app/scripts/charts/nodes-chart.js]: This is gonna be pricey.

File: client/app/scripts/components/metric-selector.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import { selectMetric } from '../actions/app-actions';
+import { MetricSelectorItem } from './metric-selector-item';
+
+// const CROSS = '\u274C';
+// const MINUS = '\u2212';
+// const DOT = '\u2022';
+//
```
Comment (davkal) [inline client/app/scripts/components/metric-selector.js]: I guess we dont need those.

File: client/app/scripts/charts/node-shape-circle.js [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
+  });
+  const metricStyle = {
+    fill: getMetricColor(metric)
+  };
+  const fontSize = size * CANVAS_METRIC_FONT_SIZE;
 
   return (
-    <g className="shape">
+    <g className={className}>
+      <defs>
```
Comment (davkal) [inline client/app/scripts/charts/node-shape-circle.js]: Could we not render this (and the metric-fill path) when no metric is selected?

File: client/app/scripts/charts/nodes-chart.js [hunk 5]
```diff
@@ -131,6 +131,13 @@ export default class NodesChart extends React.Component {
       return 1;
     };
 
+    const metric = node => {
+      const met = node.get('metrics') && node.get('metrics')
+        .filter(m => m.get('id') === this.props.selectedMetric)
+        .first();
+      return met;
+    };
+
```
Comment (foot) [inline client/app/scripts/charts/nodes-chart.js]: Pull this out and maybe pop the rest of the metrics off the nodes too, so a change in one of their values doesn't force a re-render of the chart even if its not selected.

File: client/app/scripts/stores/app-store.js [hunk 6]
```diff
@@ -564,6 +608,22 @@ export class AppStore extends Store {
...
         });
 
+        availableCanvasMetrics = nodes
+          .valueSeq()
+          .flatMap(n => (n.get('metrics') || makeList()).map(m => (
+            makeMap({id: m.get('id'), label: m.get('label')})
+          )))
+          .toSet()
+          .sortBy(m => m.get('label'))
+          .toJS();
```
Comment (davkal) [inline client/app/scripts/stores/app-store.js]: Why not leave it an immutable set/list? May help with purity later on.

File: client/app/scripts/stores/app-store.js [hunk 7]
```diff
@@ -58,6 +59,12 @@ let controlPipes = makeOrderedMap(); // pipeId -> controlPipe
 let updatePausedAt = null; // Date
 let websocketClosed = true;
 
+let selectedMetric = null;
+let pinnedMetric = selectedMetric;
+let pinnedMetricType = null;
```
Comment (davkal) [inline client/app/scripts/stores/app-store.js]: A comment would help for metric vs metricType

File: client/app/scripts/utils/metric-utils.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import _ from 'lodash';
+import d3 from 'd3';
+import { formatMetric } from './string-utils';
+import { colors } from './color-utils';
+
+
+const openFilesScale = d3.scale.log().domain([1, 100000]).range([0, 1]);
```
Comment (davkal) [inline client/app/scripts/utils/metric-utils.js]: I was hoping for these to be more generic. As in, the backed determines the type of scale and the value range. So there would be no need to hardcode `openFiles` here.

File: client/app/scripts/utils/string-utils.js [hunk 9]
```diff
@@ -4,43 +4,54 @@ import d3 from 'd3';
...
+  };
+
+  return formatters;
+}
+
+
+const formatters = makeFormatters(renderHtml);
+const svgFormatters = makeFormatters((text, unit) => `${text}${unit}`);
+
+export function formatMetric(value, opts, svg) {
```
Comment (davkal) [inline client/app/scripts/utils/string-utils.js]: Why didnt you export 2 formatMetric functions? You had to define 2 formatters (including the `svg` argument) anyway.

File: client/app/styles/main.less [hunk 10]
```diff
@@ -387,12 +387,43 @@ h2 {
     }
   }
 
-  .shape {
+  .stack .shape .highlighted {
+    display: none;
```
Comment (davkal) [inline client/app/styles/main.less]: Much better solution than `highlightOnly`.

File: client/app/styles/main.less [hunk 11]
```diff
@@ -387,12 +387,43 @@ h2 {
...
+    .node { display: none; }
+    .highlighted { display: inline; }
+  }
+
+  .stack .shape .metric-fill {
+    display: none;
+  }
+
+  .stack .onlyMetrics .shape .metric-fill {
+    display: inline-block;
```
Comment (davkal) [inline client/app/styles/main.less]: Why is that needed?

File: client/app/styles/main.less [hunk 12]
```diff
@@ -387,12 +387,43 @@ h2 {
...
+  }
 
+  .shape {
     /* cloud paths have stroke-width set dynamically */
     &:not(.shape-cloud) .border {
       stroke-width: @node-border-stroke-width;
       fill: @background-color;
+      transition: stroke-opacity 0.5s cubic-bezier(0,0,0.21,1), fill 0.5s cubic-bezier(0,0,0.21,1);
```
Comment (davkal) [inline client/app/styles/main.less]: What made you choose `cubic-bezier(0,0,0.21,1)`? In other places we use `ease-in-out`. Would be good to have transitions perform similarly to keep the *feel* the same. Maybe extract the transition style as a variable.

File: client/package.json [hunk 13]
```diff
@@ -67,7 +68,10 @@
     "test": "jest",
     "coveralls": "cat coverage/lcov.info | coveralls",
     "lint": "eslint app",
-    "clean": "rm build/app.js"
+    "clean": "rm build/app.js",
+    "noprobe": "../scope stop && ../scope launch --no-probe --app.window 24h",
+    "loadreport": "npm run noprobe && sleep 1 && curl -X POST -H \"Content-Type: application/json\" http://$BACKEND_HOST/api/report",
+    "loadreportjson": "npm run loadreport -- -d @../k8s_report.json"
```
Comment (davkal) [inline client/package.json]: I'd not hardcode a report here. Either use env variables, or allow args via `npm run loadreportjson -- @report.json`.

File: client/webpack.local.config.js [hunk 14]
```diff
@@ -26,7 +26,8 @@ module.exports = {
     'app': [
       './app/scripts/main',
       'webpack-dev-server/client?http://localhost:4041',
-      'webpack/hot/only-dev-server'
+      'webpack/hot/only-dev-server',
+      './app/scripts/debug'
```
Comment (davkal) [inline client/webpack.local.config.js]: Nice one. Let's keep that there.

**Comment by davkal** [comment]
All works well and looks good.

My personal nit would be the fill color. I feel it's too strong. Maybe use `getNodeColorDark` for more consistent coloring and try a lower opacity for a lighter feel. Or add a new `getNodeColorLight` that corrects lightness in the manner `getNodeColorDark`, just the towards consistent lightness.

File: client/app/styles/main.less (hunk 11 above)
Comment (foot) [inline client/app/styles/main.less]: Was needed for metrics on stacks, which are no more!

File: client/app/styles/main.less (hunk 12 above)
Comment (foot) [inline client/app/styles/main.less]: I c-p'd from somewhere else... the details panel transform! But variablizing is good call, for this case its not noticeable which ease we use, perhaps more so for the details-panel.

File: client/app/scripts/utils/metric-utils.js (hunk 8 above)
Comment (foot) [inline client/app/scripts/utils/metric-utils.js]: openFilesScale will disappear w/ #1149.

Load will continue w/ a vague max. Current stuff in a metric object:

```js
{
  id: "process_cpu_usage_percent",
  label: "CPU",
  format: "percent",
  value: 4.16,
  samples: [],
  min: 0,
  max: 100,
  first: "2016-04-04T08:41:49.831200363Z",
  last: "2016-04-04T08:42:02.831776457Z"
}
```

`scale: "log"`, feels a bit funny in there... but i understand the gist... I'll make an issue.

File: client/app/scripts/utils/string-utils.js (hunk 9 above)
Comment (foot) [inline client/app/scripts/utils/string-utils.js]: It was starting to get a bit too higher-order. But I'll give it a whirl. Would be nicer.
